pkg/controller/controller.go

```diff
@@ -1108,6 +1251,196 @@ func (p *csiProvisioner) volumeHandleToId(handle string) string {
+		pvc, err := p.nodeDeployment.becomeOwner(ctx, p.client, claim)
+		if err == nil && pvc == nil {
+			// Definitely not the owner. Ignore it.
+			return false, nil
+		}
+		if err != nil {
+			return false, fmt.Errorf("PVC %s/%s: %v", claim.Namespace, claim.Name, err)
+		}
+	case p.nodeDeployment.NodeName:
+		// Our node is selected.
+	default:
+		// Some other node is selected, ignore it.
+		return false, nil
+	}
+
+	// Can provision.
+	return true, nil
+}
+
+func (p *csiProvisioner) checkCapacity(ctx context.Context, claim *v1.PersistentVolumeClaim, selectedNodeName string) (bool, error) {
+	capacity := claim.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+	volSizeBytes := capacity.Value()
+	if volSizeBytes == 0 {
+		// Nothing to check.
+		return true, nil
+	}
+
+	if claim.Spec.StorageClassName == nil {
+		return false, errors.New("empty storage class name")
+	}
+	sc, err := p.scLister.Get(*claim.Spec.StorageClassName)
+	if err != nil {
+		return false, err
+	}
+
+	node, err := p.nodeLister.Get(selectedNodeName)
+	if err != nil {
+		return false, err
+	}
+
+	result, _, err := p.prepareProvision(ctx, claim, sc, node)
+	if err != nil {
+		return false, err
+	}
+
+	// In practice, we expect exactly one entry here once a node
+	// has been selected. But we have to be prepared for more than
+	// one (=> check all, success if there is at least one) and
+	// none (no node selected => check once without topology).
+	topologies := []*csi.Topology{nil}
+	if result.req.AccessibilityRequirements != nil && len(result.req.AccessibilityRequirements.Requisite) > 0 {
+		topologies = result.req.AccessibilityRequirements.Requisite
+	}
+	for _, topology := range topologies {
+		req := &csi.GetCapacityRequest{
+			VolumeCapabilities: result.req.VolumeCapabilities,
+			Parameters:         result.req.Parameters,
+			AccessibleTopology: topology,
+		}
+		klog.V(5).Infof("GetCapacityRequest %+v", req)
+		resp, err := p.csiClient.GetCapacity(ctx, req)
+		if err != nil {
+			return false, fmt.Errorf("GetCapacity: %v", err)
+		}
+		if volSizeBytes <= resp.AvailableCapacity {
+			// Enough capacity at the moment.
+			return true, nil
+		}
+	}
+
+	// Currently not enough capacity anywhere.
+	return false, nil
+}
+
+// becomeOwner updates the PVC with the current node as selected node.
+// Returns an error if something unexpectedly failed, otherwise an updated PVC with
+// the current node selected or nil if not the owner.
+func (nc *internalNodeDeployment) becomeOwner(ctx context.Context, client kubernetes.Interface, claim *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error) {
+	attempts := nc.rateLimiter.NumRequeues(claim.UID)
+	delay := nc.rateLimiter.When(claim.UID)
+	klog.V(5).Infof("will try to become owner of PVC %s/%s in %s, attempt #%d", claim.Namespace, claim.Name, delay, attempts)
+	sleep, cancel := context.WithTimeout(ctx, delay)
+	defer cancel()
+	ticker := time.NewTicker(10 * time.Millisecond)
```

Comment:
Did you consider using claimInformer to get the events?

Reply:
Instead of polling? Yes, I had been thinking about that, but couldn't come up with a good way to hook up the informer callback with this function here.

Callbacks can only be added, but not removed. So I would have to install some generic callback that gets told at runtime who is interested in which PVC - felt rather complicated for a rather small improvement in latency.

Reply:
Do we need the polling period to be so frequent considering that the delays are in seconds?

Reply:
Probably not. I've replaced it with 'base delay / 100', so now it scales together with the base delay and (for the default) is at 300ms.

I don't want to make it too large either because then one go routine would be blocked for potentially quite a while before it becomes available for some other PVC work item. With many provisioners running in parallel the actual smallest delay can be smaller than the maximum of 30 seconds.